logr.go

```diff
@@ -198,16 +198,18 @@ import (
 // New returns a new Logger instance.  This is primarily used by libraries
 // implementing LogSink, rather than end users.
 func New(sink LogSink) Logger {
-	logger := Logger{
-		Sink: sink,
-	}
-	if withCallDepth, ok := sink.(CallDepthLogSink); ok {
-		logger.withCallDepth = withCallDepth
-	}
+	logger := Logger{}
+	logger.SetSink(sink)
 	sink.Init(runtimeInfo)
 	return logger
 }
 
+// SetSink stores the sink and updates all related fields.
+func (l *Logger) SetSink(sink LogSink) {
+	l.Sink = sink
+	l.withCallDepth, _ = l.Sink.(CallDepthLogSink)
```

Comment:
This isn't right, obviously.  We'd need to test the interface.

Reply:
Is this not guaranteed to return nil if the check fails?

Reply:
The rationale behind always setting l.withCallDepth was that, theoretically at least, l.withCallDepth might be set from a previous call and then l.Sink gets replaced by a sink that no longer implements WithCallDepth. I don't know whether that is realistic, but I'd rather not make such assumptions.

The alternative code is:
```
if withCallDepth, ok := l.Sink.(CallDepthLogSink); ok {
    l.withCallDepth = withCallDepth
} else {
    l.withCallDepth = nil
}
```

Reply:
I missed that the 2nd return, even when caught and discarded, inhibits the panic.